### PR TITLE
Submit KPI data when the dialog closes either in the interaction_iframe OR in the internal_api.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
@kparlante - this one is for you!

Only generate GUIDs for a KPI blob whenever needed:
- When the dialog is orphaned AND
- A staging event occurs without a confirmation event.

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
fixes #3306

File: resources/static/common/js/models/interaction_data.js [hunk 1]
```diff
@@ -80,6 +80,26 @@ BrowserID.Models.InteractionData = (function() {
     }
   }
 
+  function publishCurrent(oncomplete) {
+    stageCurrent();
+    publishStaged(oncomplete);
+  }
+
+  function indexOfEvent(eventStream, eventName) {
```
Comment (shane-tomlinson) [inline resources/static/common/js/models/interaction_data.js:142]: These two functions are no longer needed with the removal of the linking_id

**Comment by kparlante** [comment]
@shane-tomlinson : I played with this one and did indeed see the blobs on the first go round (yay!). I tried with both myfavoritebeer and 123done, covering both get and watch apis.

While I don't I need a matching guid to link blobs when the user closes the dialog to wait for email verification, I do need some sort of flag for that next blob that succeeds, so I can tally a count for eventual new user successes.

FWIW, we can get rid of the orphaned flag if its only meaning is the existence of an assertion generated event.

**Comment by shane-tomlinson** [comment]
@kparlante - What is the metric needs to be captured, % of signups that eventually complete? Something else? 

I'm shooting in the dark here, but I am trying to figure out how to handle the following scenarios:
1. User opens dialog, stages nothing, closes dialog.
2. User stages an address, closes the dialog, never does anything
3. User stages an address, verifies in a new tab while dialog is still open.
4. User stages an address, closes the dialog, and then verifies in same browser.
5. User stages an address, verifies in a different browser while the dialog is still open.
6. User stages an address, closes the dialog, verifies in a different browser.

#### Walking through cases

##### 1. User opens dialog, stages nothing, closes dialog.

KPI data sent when user closes dialog.

##### 2. User stages an address, closes the dialog, never does anything

KPI data sent when user closes dialog.

##### 3. User stages an address, verifies in a new tab while dialog is still open.

KPI data sent when dialog closes post-verification. KPI blob's event stream indicates that verification has completed and an assertion has been generated.

##### 4. User stages an address, closes the dialog, and then verifies in same browser.

KPI data is sent when user closes dialog, pre-verification. KPI blob's event stream indicates an address has been staged, but not that one has completed.

##### 5. User stages an address, verifies in a different browser while the dialog is still open.

KPI data is sent when dialog closes post-verification, like case 3. No KPI blob sent from second browser.

##### 6. User stages an address, closes the dialog, verifies in a different browser.

KPI data is sent when user closes dialog, pre-verification. No KPI blob sent from second browser.

Cases 3-6 are successful and need to indicate "success!"

In cases 3 and 5, we could use the "assertion_generated" event to indicate success, but for cases 4 & 6, since the user has already closed the dialog and they are not guaranteed to open the dialog a second time, we can't rely on using assertion_generated the next time the dialog is open to give us this data.

We currently do not log any events for wsapi calls on the main site, meaning no event is ever added to the event stream when an address is verified. A verification event seems like success criteria. 

If so, should we start logging events on email verification pages to capture successes?

**Comment by kparlante** [comment]
@shane-tomlinson : We're trying to capture how many stages each "new user" makes it through when they come to the site. So, ideally, for a given day + segment, we want # "new" users @ screen.set_password, # "new" users @ user.user_staged, # "new" users @ user.user_confirmed, # "new" users @ assertion_generated. (We should probably add number of users who don't make it to screen.set_password as well, although we don't know for sure that they are "new" users).

I should add that its not important to be perfect, we just need the data to be good enough that we trust we have a pretty good picture of what's going on.

We can handle/count cases 1, 2, and 3 easily now, just by looking at the stages (especially with this fix where we actually get all of the blobs). We also easily handle case 5, assuming we see the assertion_generated in the event stream.

We were using "orphaned" to handle case 4, because the original blob would get modified to orphaned = false. I could then add counts for the last two stages if I saw orphaned = false + only the first two stages. This is the case that gets lost with the new change. I was thinking we could add a flag to the second blob that gets generated, something that means "this blob completes some previous blob, even if we're not telling you _which_ previous blob". I can then add counts to the last two stages for that segment/day. Basically a flag in place of a guid.

We're just losing case 6. I suppose we could handle it if they _ever_ come back to that original browser, I assume that ends up looking like case 4.

@lloyd and I talked about your idea of logging email verifications on the main site yesterday -- yes, I think this would be helpful to cross check the data we see in the KPIs. We could estimate case 6 for example.

**Comment by shane-tomlinson** [comment]
@kparlante thanks for the explanation. I imagine any new flag would be cause for a data privacy review, so I am wondering if we can do this using the event stream. Instead of adding a new flag could we cover # 4 and 6 by logging verifications? I'm not exactly sure how it would look/work yet, I am trying to think up alternatives.

**Comment by kparlante** [comment]
@shane-tomlinson Well, the first item on the event stream could be "removed-previous-orphaned-blob" or similar :-p .
Logging verifications will similarly require review (I suspect we'll want to do a review at the beginning of next quarter for a handful of things we want to do).

I think its useful to go ahead with this change even if we don't cover cases #4 and #6, so I do +r the change. If we go ahead without addressing it we should just log a new issue.

Another argument for not needing a review... isn't it logically the same as de-orphaning a blob?

File: resources/static/dialog/js/misc/internal_api.js [hunk 2]
```diff
@@ -178,6 +177,18 @@
     }, complete.curry(null));
   }
 
+  function setOrigin(origin) {
```
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/internal_api.js:180]: This code is just moved from above for better proximity.

**Comment by kparlante** [comment]
@shane-tomlinson : excellent! I tested this out for cases 4 and 6 (or at least the variant of 6 where the user eventually comes back to the original browser), works nicely. +r

**Pull request merged**